README.md
HExcel
------

[![Hackage](https://img.shields.io/hackage/v/HExcel.svg?style=flat)](https://hackage.haskell.org/package/HExcel)


## Create Excel files with Haskell


This is a fork of [libxlsxwriter](https://github.com/HalfWayMan/libxlsxwriter)
that tries the improve on the api and provide a library for
creation of Excel files.
Underneath the hood it uses C library called [libxlsxwriter](http://libxlsxwriter.github.io/) and provides
bindings to C code to produce Excel 2007+ xlsx files.

#### Example

```
{-# LANGUAGE TypeApplications #-}
module Main where

import Control.Monad.Trans.State (execStateT)
import Control.Monad (forM_)
import Data.Time (getZonedTime)
import HExcel

main :: IO ()
main = do
  wb <- workbookNew "test.xlsx"
  let props = mkDocProperties
        { docPropertiesTitle   = "Test Workbook"
        , docPropertiesCompany = "HExcel"
        }
  workbookSetProperties wb props
  ws <- workbookAddWorksheet wb "First Sheet"
  df <- workbookAddFormat wb
  formatSetNumFormat df "mmm d yyyy hh:mm AM/PM"
  now <- getZonedTime
  -- You can create HExcelState which is convenient api for writing to cells 
  let initState = HExcelState Nothing ws 4 1 0 1 0
  _ <- flip execStateT initState $ do
         writeCell "David"
         writeCell "Dimitrije"
		 -- we can skip some rows
         skipRows 1
         writeCell "Jovana"
		 -- or skip some columns
         skipCols 1
         writeCell (zonedTimeToDateTime now)
         writeCell @Double 42.5

  -- or use functions that run in plain IO
  forM_ [5 .. 8] $ \n -> do
    writeString ws Nothing n 3 "xxx"
    writeNumber ws Nothing n 4 1234.56
    writeDateTime ws (Just df) n 5 (zonedTimeToDateTime now)
  workbookClose wb

```


CHANGELOG.md
# Revision history for HExcel

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


src/HExcel/Ffi.hs
{-# LANGUAGE ForeignFunctionInterface #-}

-- |
-- Module      :  HExcel.Ffi
-- Maintainer  :  Sasha Bogicevic <sasa.bogicevic@pm.me>
-- Stability   :  experimental
--
-- FFI code

module HExcel.Ffi where

import Foreign
import Foreign.C.String
import HExcel.Types

foreign import ccall "workbook_new"
  workbook_new :: CString -> IO (Ptr LxwWorkbook_)

foreign import ccall "workbook_new_opt"
  workbook_new_opt :: CString -> Ptr WorkbookOptions -> IO (Ptr LxwWorkbook_)

foreign import ccall "workbook_close"
  workbook_close :: Ptr LxwWorkbook_ -> IO ()

foreign import ccall "workbook_add_worksheet"
  workbook_add_worksheet :: Ptr LxwWorkbook_ -> CString ->
                            IO (Ptr LxwWorksheet_)

foreign import ccall "workbook_add_format"
  workbook_add_format :: Ptr LxwWorkbook_ -> IO (Ptr LxwFormat_)

foreign import ccall "workbook_set_properties"
  workbook_set_properties :: Ptr LxwWorkbook_ ->
                             Ptr DocProperties' -> IO ()

foreign import ccall "workbook_define_name"
  workbook_define_name :: Ptr LxwWorkbook_ -> CString -> CString -> IO ()

foreign import ccall "worksheet_write_number"
  worksheet_write_number :: Ptr LxwWorksheet_ ->
                            Word32 -> Word16 ->
                            Double -> Ptr LxwFormat_ -> IO ()

foreign import ccall "worksheet_write_string"
  worksheet_write_string :: Ptr LxwWorksheet_ ->
                            Word32 -> Word16 ->
                            CString -> Ptr LxwFormat_ -> IO ()

foreign import ccall "worksheet_write_formula"
  worksheet_write_formula :: Ptr LxwWorksheet_ ->
                             Word32 -> Word16 ->
                             CString -> Ptr LxwFormat_ -> IO ()

foreign import ccall "worksheet_write_array_formula"
  worksheet_write_array_formula :: Ptr LxwWorksheet_ ->
                                   Word32 -> Word16 ->
                                   Word32 -> Word16 ->
                                   CString -> Ptr LxwFormat_ -> IO ()

foreign import ccall "worksheet_write_datetime"
  worksheet_write_datetime :: Ptr LxwWorksheet_ ->
                              Word32 -> Word16 ->
                              Ptr DateTime -> Ptr LxwFormat_ -> IO ()

foreign import ccall "worksheet_write_url"
  worksheet_write_url :: Ptr LxwWorksheet_ ->
                         Word32 -> Word16 ->
                         CString -> Ptr LxwFormat_ -> IO ()

foreign import ccall "worksheet_set_row"
  worksheet_set_row :: Ptr LxwWorksheet_ ->
                       Word32 -> Double -> Ptr LxwFormat_ -> IO ()

foreign import ccall "worksheet_set_column"
  worksheet_set_column :: Ptr LxwWorksheet_ ->
                          Word16 -> Word16 -> Double -> Ptr LxwFormat_ -> IO ()

foreign import ccall "worksheet_insert_image"
  worksheet_insert_image :: Ptr LxwWorksheet_ ->
                            Word32 -> Word16 -> CString -> IO ()

foreign import ccall "worksheet_insert_image_opt"
  worksheet_insert_image_opt :: Ptr LxwWorksheet_ ->
                                Word32 -> Word16 ->
                                CString -> Ptr ImageOptions -> IO ()

foreign import ccall "worksheet_merge_range"
  worksheet_merge_range :: Ptr LxwWorksheet_ ->
                           Word32 -> Word16 ->
                           Word32 -> Word16 ->
                           CString -> Ptr LxwFormat_ -> IO ()

foreign import ccall "worksheet_freeze_panes"
  worksheet_freeze_panes :: Ptr LxwWorksheet_ ->
                            Word32 -> Word16 -> IO ()

foreign import ccall "worksheet_split_panes"
  worksheet_split_panes :: Ptr LxwWorksheet_ ->
                           Double -> Double -> IO ()

foreign import ccall "worksheet_set_landscape"
  worksheet_set_landscape :: Ptr LxwWorksheet_ -> IO ()

foreign import ccall "worksheet_set_portrait"
  worksheet_set_portrait :: Ptr LxwWorksheet_ -> IO ()

foreign import ccall "worksheet_set_page_view"
  worksheet_set_page_view :: Ptr LxwWorksheet_ -> IO ()

foreign import ccall "worksheet_set_paper"
  worksheet_set_paper :: Ptr LxwWorksheet_ -> Word8 -> IO ()

foreign import ccall "worksheet_set_margins"
  worksheet_set_margins :: Ptr LxwWorksheet_ ->
                           Double -> Double ->
                           Double -> Double -> IO ()

foreign import ccall "worksheet_set_header"
  worksheet_set_header :: Ptr LxwWorksheet_ -> CString -> IO ()

foreign import ccall "worksheet_set_footer"
  worksheet_set_footer :: Ptr LxwWorksheet_ -> CString -> IO ()

foreign import ccall "worksheet_set_zoom"
  worksheet_set_zoom :: Ptr LxwWorksheet_ -> Word16 -> IO ()

foreign import ccall "worksheet_set_print_scale"
  worksheet_set_print_scale:: Ptr LxwWorksheet_ -> Word16 -> IO ()

foreign import ccall "format_set_font_name"
  format_set_font_name :: Ptr LxwFormat_ -> CString -> IO ()

foreign import ccall "format_set_font_size"
  format_set_font_size :: Ptr LxwFormat_ -> Word16 -> IO ()

foreign import ccall "format_set_font_color"
  format_set_font_color :: Ptr LxwFormat_ -> Int32 -> IO ()

foreign import ccall "format_set_num_format"
  format_set_num_format :: Ptr LxwFormat_ -> CString -> IO ()

foreign import ccall "format_set_bold"
  format_set_bold :: Ptr LxwFormat_ -> IO ()

foreign import ccall "format_set_italic"
  format_set_italic :: Ptr LxwFormat_ -> IO ()

foreign import ccall "format_set_underline"
  format_set_underline :: Ptr LxwFormat_ -> Word8 -> IO ()

foreign import ccall "format_set_font_strikeout"
  format_set_font_strikeout :: Ptr LxwFormat_ -> IO ()

foreign import ccall "format_set_font_script"
  format_set_font_script :: Ptr LxwFormat_ -> Word8 -> IO ()

foreign import ccall "format_set_num_format_index"
  format_set_num_format_index :: Ptr LxwFormat_ -> Word8 -> IO ()

foreign import ccall "format_set_align"
  format_set_align :: Ptr LxwFormat_ -> Word8 -> IO ()

foreign import ccall "format_set_text_wrap"
  format_set_text_wrap :: Ptr LxwFormat_ -> IO ()

foreign import ccall "format_set_rotation"
  format_set_rotation :: Ptr LxwFormat_ -> Int16 -> IO ()

foreign import ccall "format_set_shrink"
  format_set_shrink :: Ptr LxwFormat_ -> IO ()

foreign import ccall "format_set_pattern"
  format_set_pattern :: Ptr LxwFormat_ -> Word8 -> IO ()

foreign import ccall "format_set_bg_color"
  format_set_bg_color :: Ptr LxwFormat_ -> Int32 -> IO ()

foreign import ccall "format_set_fg_color"
  format_set_fg_color :: Ptr LxwFormat_ -> Int32 -> IO ()

foreign import ccall "format_set_border"
  format_set_border :: Ptr LxwFormat_ -> Word8 -> IO ()

foreign import ccall "format_set_bottom"
  format_set_bottom :: Ptr LxwFormat_ -> Word8 -> IO ()

foreign import ccall "format_set_top"
  format_set_top :: Ptr LxwFormat_ -> Word8 -> IO ()

foreign import ccall "format_set_left"
  format_set_left :: Ptr LxwFormat_ -> Word8 -> IO ()

foreign import ccall "format_set_right"
  format_set_right :: Ptr LxwFormat_ -> Word8 -> IO ()

foreign import ccall "format_set_border_color"
  format_set_border_color :: Ptr LxwFormat_ -> Int32 -> IO ()

foreign import ccall "format_set_bottom_color"
  format_set_bottom_color :: Ptr LxwFormat_ -> Int32 -> IO ()

foreign import ccall "format_set_top_color"
  format_set_top_color :: Ptr LxwFormat_ -> Int32 -> IO ()

foreign import ccall "format_set_left_color"
  format_set_left_color :: Ptr LxwFormat_ -> Int32 -> IO ()

foreign import ccall "format_set_right_color"
  format_set_right_color :: Ptr LxwFormat_ -> Int32 -> IO ()


src/HExcel/Types.hs
{-# LANGUAGE TemplateHaskell #-}

-- |
-- Module      :  HExcel.Types
-- Maintainer  :  Sasha Bogicevic <sasa.bogicevic@pm.me>
-- Stability   :  experimental
--
-- HExcel types are defined here
module HExcel.Types where

import Data.Time
import Data.Word
import Foreign
import Foreign.C.String
import Foreign.C.Types
import Lens.Micro.TH

-- | Excel Row
type Row = Word32
-- | Excel Column
type Col = Word16

-- | Pointer to Excel Workbook
data LxwWorkbook_
-- | Excel Workbook
newtype Workbook = Workbook (Ptr LxwWorkbook_)

data LxwWorksheet_
-- | Excel WorkSheet
newtype Worksheet = Worksheet (Ptr LxwWorksheet_)

-- | Pointer to Excel Format
data LxwFormat_
-- | Excel Format
newtype Format = Format { unFormat :: Ptr LxwFormat_ }

-- | 'HExcelState' is the state we thread trough for the writeCell function of the
-- HExcel typeclass. We are trying to create a convenient api for writing cell values
-- without too much hassle.
data HExcelState =
  HExcelState
    { _hexcelStateFormat     :: Maybe Format
      -- ^ define a formatting to be used on all cells
    , _hexcelStateSheet      :: Worksheet
      -- ^ define a worksheet to write to
    , _hexcelStateRowCeiling :: Word32
      -- ^ max row size
    , _hexcelStateInitRow    :: Row
      -- ^ first row to write to
    , _hexcelStateInitCol    :: Col
      -- ^ first column to write to
    , _hexcelStateRow        :: Row
      -- ^ current row
    , _hexcelStateCol        :: Col
      -- ^ current column
    }

makeLenses ''HExcelState

-- | Excel WorkBook Options
newtype WorkbookOptions =
  WorkbookOptions { workbookOptionsConstantMem :: Bool }

instance Storable WorkbookOptions where
  sizeOf _ = sizeOf (undefined :: Word8)
  alignment _ = alignment (undefined :: Int)
  peek ptr = do
    cm <- peekByteOff ptr 0 :: IO Word8
    return (WorkbookOptions (cm /= 1))
  poke ptr opts = do
    let cm | workbookOptionsConstantMem opts = 1
           | otherwise = 0
    pokeByteOff ptr 0 (cm :: Word8)

-- | Colors
data Color
  = ColorBlack
  | ColorBlue
  | ColorBrown
  | ColorCyan
  | ColorGray
  | ColorGreen
  | ColorLime
  | ColorMagenta
  | ColorNavy
  | ColorOrange
  | ColorPink
  | ColorPurple
  | ColorRed
  | ColorSilver
  | ColorWhite
  | ColorYellow
  | Color Word8 Word8 Word8

-- | Underline styles
data UnderlineStyle
  = UnderlineNone
  | UnderlineSingle
  | UnderlineDouble
  | UnderlineSingleAccounting
  | UnderlineDoubleAccounting
  deriving (Eq, Enum, Read, Show)

-- | Script styles
data ScriptStyle
  = SuperScript
  | SubScript
  deriving (Eq, Enum, Read, Show)

-- | Alignment styles
data Align
  = AlignNone
  | AlignLeft
  | AlignCenter
  | AlignRight
  | AlignFill
  | AlignJustify
  | AlignCenterAcross
  | AlignDistributed
  deriving (Eq, Enum, Read, Show)

-- | Vertical align styles
data VerticalAlign
  = VerticalAlignNone
  | VerticalAlignTop
  | VerticalAlignBottom
  | VerticalAlignCenter
  | VerticalAlignJustify
  | VerticalAlignDistributed
  deriving (Eq, Enum, Read, Show)

-- | Pattern styles
data Pattern
  = PatternNone
  | PatternSolid
  | PatternMediumGray
  | PatternDarkGray
  | PatternLightGray
  | PatternDarkHorizontal
  | PatternDarkVertical
  | PatternDarkDown
  | PatternDarkUp
  | PatternDarkGrid
  | PatternDarkTrellis
  | PatternLightHorizontal
  | PatternLightVertical
  | PatternLightDown
  | PatternLightUp
  | PatternLightGrid
  | PatternLightTrellis
  | PatternGray125
  | PatternGray0625
  deriving (Eq, Enum, Read, Show)

-- | Border options
data Border
  = BorderAll
  | BorderBottom
  | BorderTop
  | BorderLeft
  | BorderRight
  deriving (Eq, Read, Show)

-- | Border styles
data BorderStyle
  = BorderNone
  | BorderThin
  | BorderMedium
  | BorderDashed
  | BorderDotted
  | BorderThick
  | BorderDouble
  | BorderHair
  | BorderMediumDashed
  | BorderDashDot
  | BorderMediumDashDot
  | BorderDashDotDot
  | BorderMediumDashDotDot
  | BorderSlantDashDot
  deriving (Eq, Enum, Read, Show)

-- | Excel document properties
data DocProperties =
  DocProperties { docPropertiesTitle         :: String
                , docPropertiesSubject       :: String
                , docPropertiesAuthor        :: String
                , docPropertiesManager       :: String
                , docPropertiesCompany       :: String
                , docPropertiesCategory      :: String
                , docPropertiesKeywords      :: String
                , docPropertiesComments      :: String
                , docPropertiesStatus        :: String
                , docPropertiesHyperlinkBase :: String
                , docPropertiesCreated       :: UTCTime {-CTime-}
               }

data DocProperties' =
  DocProperties' { docPropsTitle         :: CString
                 , docPropsSubject       :: CString
                 , docPropsAuthor        :: CString
                 , docPropsManager       :: CString
                 , docPropsCompany       :: CString
                 , docPropsCategory      :: CString
                 , docPropsKeywords      :: CString
                 , docPropsComments      :: CString
                 , docPropsStatus        :: CString
                 , docPropsHyperlinkBase :: CString
                 , docPropsCreated       :: CTime
                 }

instance Storable DocProperties' where
  sizeOf _ = 10 * sizeOf (undefined :: CString) +
                  sizeOf (undefined :: CTime)
  alignment _ = alignment (undefined :: CString)
  peek = error "No implementation of 'peek' for 'DocProperties'"
  poke ptr props = do
    let n = sizeOf (undefined :: CString)
    pokeByteOff ptr (0 * n) (docPropsTitle props)
    pokeByteOff ptr (1 * n) (docPropsSubject props)
    pokeByteOff ptr (2 * n) (docPropsAuthor props)
    pokeByteOff ptr (3 * n) (docPropsManager props)
    pokeByteOff ptr (4 * n) (docPropsCompany props)
    pokeByteOff ptr (5 * n) (docPropsCategory props)
    pokeByteOff ptr (6 * n) (docPropsKeywords props)
    pokeByteOff ptr (7 * n) (docPropsComments props)
    pokeByteOff ptr (8 * n) (docPropsStatus props)
    pokeByteOff ptr (9 * n) (docPropsHyperlinkBase props)
    pokeByteOff ptr (10 * n) (docPropsCreated props)

-- | Type to hold datetime values
data DateTime =
  DateTime { dtYear   :: CInt
           , dtMonth  :: CInt
           , dtDay    :: CInt
           , dtHour   :: CInt
           , dtMinute :: CInt
           , dtSecond :: CDouble
           }
  deriving (Show)

instance Storable DateTime where
  sizeOf _ = 5 * sizeOf (undefined :: Int) +
                 sizeOf (undefined :: Double)
  alignment _ = alignment (undefined :: Int)
  peek ptr = do
    let ptr' = castPtr ptr
    DateTime <$> peekElemOff ptr' 0
             <*> peekElemOff ptr' 1
             <*> peekElemOff ptr' 2
             <*> peekElemOff ptr' 3
             <*> peekElemOff ptr' 4
             <*> peekElemOff ptr' 5
  poke ptr (DateTime y m d h mi s) = do
    pokeByteOff ptr 0 y
    pokeByteOff ptr 4 m
    pokeByteOff ptr 8 d
    pokeByteOff ptr 12 h
    pokeByteOff ptr 16 mi
    pokeByteOff ptr 20 s

-- | Type to hold image options
data ImageOptions =
  ImageOptions { imageOffsetX :: Int32
               , imageOffsetY :: Int32
               , imageScaleX  :: Double
               , imageScaleY  :: Double
               }

instance Storable ImageOptions where
  sizeOf _ = 2 * sizeOf (undefined :: Int32) +
             2 * sizeOf (undefined :: Double)
  alignment _ = alignment (undefined :: Int32)
  peek ptr =
    ImageOptions <$> peekByteOff ptr 0
                 <*> peekByteOff ptr 4
                 <*> peekByteOff ptr 8
                 <*> peekByteOff ptr 16
  poke ptr (ImageOptions ox oy sx sy) = do
    pokeByteOff ptr 0 ox
    pokeByteOff ptr 4 oy
    pokeByteOff ptr 8 sx
    pokeByteOff ptr 16 sy

-- | Paper size
data PaperSize
  = DefaultPaper
  | LetterPaper
  | A3Paper
  | A4Paper
  | A5Paper
  | OtherPaper Word8
  deriving (Eq)


src/HExcel.hs
-- |
-- Module      :  HExcel
-- Maintainer  :  Sasha Bogicevic <sasa.bogicevic@pm.me>
-- Stability   :  experimental
--
-- Module exports
--
-- Example usage:
--
-- > {-# LANGUAGE TypeApplications #-}
-- > module Main where
-- >
-- > import Control.Monad.Trans.State (execStateT)
-- > import Control.Monad (forM_)
-- > import Data.Time (getZonedTime)
-- > import HExcel
-- >
-- > main :: IO ()
-- > main = do
-- >   wb <- workbookNew "test.xlsx"
-- >   let props = mkDocProperties
-- >         { docPropertiesTitle   = "Test Workbook"
-- >         , docPropertiesCompany = "HExcel"
-- >         }
-- >   workbookSetProperties wb props
-- >   ws <- workbookAddWorksheet wb "First Sheet"
-- >   df <- workbookAddFormat wb
-- >   formatSetNumFormat df "mmm d yyyy hh:mm AM/PM"
-- >   now <- getZonedTime
-- >   -- You can create HExcelState which is convenient api for writing to cells
-- >   let initState = HExcelState Nothing ws 4 1 0 1 0
-- >   _ <- flip execStateT initState $ do
-- >          writeCell "David"
-- >          writeCell "Dimitrije"
-- >          -- we can skip some rows
-- >          skipRows 1
-- >          writeCell "Jovana"
-- >          -- skip some columns
-- >          skipCols 1
-- >          writeCell (zonedTimeToDateTime now)
-- >          writeCell @Double 42.5
-- >
-- >   -- or use functions that run in plain IO
-- >   forM_ [5 .. 8] $ \n -> do
-- >     writeString ws Nothing n 3 "xxx"
-- >     writeNumber ws Nothing n 4 1234.56
-- >     writeDateTime ws (Just df) n 5 (zonedTimeToDateTime now)
-- >   workbookClose wb

module HExcel
  ( module HExcel.HExcelInternal
  )
  where

import HExcel.HExcelInternal
